app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java

```diff
@@ -0,0 +1,375 @@
+import android.net.Uri;
+import android.preference.PreferenceManager;
+import android.provider.Settings;
+import android.util.Log;
+import android.view.View;
+import android.widget.Toast;
+
+import androidx.annotation.NonNull;
+import androidx.core.app.ActivityCompat;
+import androidx.lifecycle.LiveData;
+
+import com.google.android.gms.location.Geofence;
+import com.google.android.gms.location.GeofencingClient;
+import com.google.android.gms.location.GeofencingRequest;
+import com.google.android.gms.location.LocationServices;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+import com.google.android.material.snackbar.Snackbar;
+import com.gusta.wakemehome.BuildConfig;
+import com.gusta.wakemehome.R;
+import com.gusta.wakemehome.utilities.Constants;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.gusta.wakemehome.utilities.Constants.ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE;
+
+public class GeofenceManager implements OnCompleteListener<Void> {
+
+    //===========//
+    // CONSTANTS //
+    //===========//
+
+    // Constant for logging
+    private static final String TAG = GeofenceManager.class.getSimpleName();
+
+    //=======//
+    // ENUMS //
+    //=======//
+
+    /**
+     * Tracks whether the user requested to add or remove geofences, or to do neither.
+     */
+    private enum PendingGeofenceTask {
+        REFRESH, NONE
+    }
+
+    //=========//
+    // MEMBERS //
+    //=========//
+
+    /**
+     * The activity so show messages (like errors) on.
+     */
+    private Activity mActivity;
+    /**
+     * The list of entries tracked - each entry will need a geofence.
+     */
+    private LiveData<? extends List<? extends GeofenceEntry>> mLiveData;
+    /**
+     * Provides access to the Geofencing API.
+     */
+    private GeofencingClient mGeofencingClient;
+    /**
+     * Used when requesting to add or remove geofences.
+     */
+    private PendingIntent mGeofencePendingIntent;
+
+    private PendingGeofenceTask mPendingGeofenceTask = PendingGeofenceTask.NONE;
+
+    //=========//
+    // METHODS //
+    //=========//
+
+    public GeofenceManager(Activity activity, LiveData<? extends List<? extends GeofenceEntry>> liveData) {
+        mActivity = activity;
+        mLiveData = liveData;
+        mGeofencingClient = LocationServices.getGeofencingClient(activity);
+    }
+
+    public void refresh() {
+        // Ask for location permissions if not yet granted
+        if (!checkPermissions()) {
+            mPendingGeofenceTask = PendingGeofenceTask.REFRESH;
+            requestPermissions();
+            return;
+        }
+
+        // TODO: remove only irrelevant geofences and add only the new ones
+        // Remove all geofences and back all the relevant ones
+        removeGeofences();
```

Comment:
Does it possible to add without granted?

Reply:
No, it fails